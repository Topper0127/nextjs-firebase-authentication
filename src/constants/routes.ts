export const SIGN_IN = '/sign-in';
export const SIGN_UP = '/sign-up';
export const ACCOUNT = '/account';
export const PASSWORD_CHANGE = '/password-change';
export const PASSWORD_FORGOT = '/password-forgot';
export const EMAIL_CHANGE = '/email-change';
export const PARTNER = '/partner';
export const COMMUNITY_JOIN = '/community-join';

export const INDEX = '/';
export const UNLOCKED_COURSE_DETAILS = '/p/[unlocked-course-id]';

export const COURSE_UPGRADE = '/upgrade/[upgradeable-course-id]';

export const CHECKOUT = '/checkout';
