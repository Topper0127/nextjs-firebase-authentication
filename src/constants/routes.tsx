export const INDEX = '/';
export const SIGN_IN = '/sign-in';
export const SIGN_UP = '/sign-up';
export const ACCOUNT = '/account';
export const PASSWORD_CHANGE = '/password-change';
export const PASSWORD_FORGOT = '/password-forgot';
export const REFERRAL = '/referral';
