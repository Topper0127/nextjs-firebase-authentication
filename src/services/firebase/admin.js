import * as firebaseAdmin from 'firebase-admin';

import firebaseServiceAccountKey from '../../../firebaseServiceAccountKey.json';

if (!firebaseAdmin.apps.length) {
  firebaseAdmin.initializeApp({
    credential: firebaseAdmin.credential.cert(
      firebaseServiceAccountKey
    ),
    databaseURL: process.env.FIREBASE_DATABASE_URL,
  });
}

export default firebaseAdmin;
