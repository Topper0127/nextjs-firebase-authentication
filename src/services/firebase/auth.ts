export const doPasswordReset = (email: string) => {
  // auth.sendPasswordResetEmail(email);
};

export const doPasswordUpdate = (password: string) => {
  // if (auth.currentUser) {
  //   return auth.currentUser.updatePassword(password);
  // } else {
  //   return Promise.reject(new Error('No authenticated user.'));
  // }
};
