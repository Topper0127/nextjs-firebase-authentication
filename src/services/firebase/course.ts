import * as firebaseAdminVanilla from 'firebase-admin';

import firebaseAdmin from '@services/firebase/admin';

import { COURSE } from '@data/course-keys-types';
import { BUNDLE } from '@data/bundle-keys-types';

export const createCourse = async ({
  uid,
  courseId,
  bundleId,
  amount,
  paymentType,
  coupon,
}: {
  uid?: string;
  courseId: COURSE;
  bundleId: BUNDLE;
  amount: number;
  paymentType: string;
  coupon: string;
}) =>
  await firebaseAdmin
    .database()
    .ref(`users/${uid}/courses`)
    .push()
    .set({
      courseId: courseId,
      packageId: bundleId,
      invoice: {
        createdAt:
          firebaseAdminVanilla.database.ServerValue.TIMESTAMP,
        amount,
        licensesCount: 1,
        currency: 'USD',
        paymentType,
        coupon,
      },
    });
