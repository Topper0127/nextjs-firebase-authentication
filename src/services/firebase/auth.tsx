import { auth } from './firebase';

export const doCreateUserWithEmailAndPassword = (
  email: string,
  password: string
) => auth.createUserWithEmailAndPassword(email, password);

export const doSignInWithEmailAndPassword = (
  email: string,
  password: string
) => auth.signInWithEmailAndPassword(email, password);

export const doSignOut = () => auth.signOut();

export const doPasswordReset = (email: string) =>
  auth.sendPasswordResetEmail(email);

export const doPasswordUpdate = (password: string) => {
  if (auth.currentUser) {
    auth.currentUser.updatePassword(password);
  }
};
