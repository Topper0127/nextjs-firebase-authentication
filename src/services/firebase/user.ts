import { db } from './firebase';

export const doCreateUser = (
  id: string,
  username: string,
  email: string
) =>
  db.ref(`users/${id}`).set({
    username,
    email,
  });
