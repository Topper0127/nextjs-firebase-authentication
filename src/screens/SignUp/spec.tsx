import renderer from 'react-test-renderer';
import waitForExpect from 'wait-for-expect';

import { message } from 'antd';

import * as ROUTES from '@constants/routes';
import * as authService from '@services/firebase/auth';

import SignUpPage from '.';

describe('SignUpPage', () => {
  let component: any;
  let getComponent: any;
  let defaultProps: any;

  beforeEach(() => {
    defaultProps = {};

    getComponent = (props: any = defaultProps) => (
      <SignUpPage {...props} />
    );

    component = renderer.create(getComponent());
  });

  beforeEach(() => {
    message.loading = jest.fn();
    message.error = jest.fn();
    message.success = jest.fn();
  });

  it('renders', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a sign in link', () => {
    expect(
      component.root.findByProps({ href: ROUTES.SIGN_IN })
    ).toBeDefined();
  });

  describe('signs up a user', () => {
    const username = 'example';
    const email = 'example@example.com';
    const password = 'mypassword';

    beforeEach(() => {
      component.root
        .findByProps({ id: '/sign-up_username' })
        .props.onChange({ target: { value: username } });

      component.root
        .findByProps({ id: '/sign-up_email' })
        .props.onChange({ target: { value: email } });

      component.root
        .findByProps({ id: '/sign-up_password' })
        .props.onChange({
          target: { value: password },
        });

      component.root
        .findByProps({ id: '/sign-up_confirm' })
        .props.onChange({
          target: { value: password },
        });
    });

    it('with success', async () => {
      const spy = jest
        .spyOn(authService, 'doCreateUserWithEmailAndPassword')
        .mockImplementation(() =>
          Promise.resolve({
            credential: null,
            user: null,
          })
        );

      component.root
        .findByType('form')
        .props.onSubmit({ preventDefault: () => {} });

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(email, password);

      expect(message.loading).toHaveBeenCalledTimes(1);

      await waitForExpect(() => {
        expect(message.error).toHaveBeenCalledTimes(0);
        expect(message.success).toHaveBeenCalledTimes(1);
      });
    });

    it('with error', async () => {
      const spy = jest
        .spyOn(authService, 'doCreateUserWithEmailAndPassword')
        .mockImplementation(() =>
          Promise.reject(new Error('Weak password.'))
        );

      component.root
        .findByType('form')
        .props.onSubmit({ preventDefault: () => {} });

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(email, password);

      expect(message.loading).toHaveBeenCalledTimes(1);

      await waitForExpect(() => {
        expect(message.error).toHaveBeenCalledTimes(1);
        expect(message.success).toHaveBeenCalledTimes(0);
      });
    });
  });
});
