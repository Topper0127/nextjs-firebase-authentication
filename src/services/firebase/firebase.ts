import app from 'firebase/app';

import 'firebase/database';
import 'firebase/auth';
import 'firebase/storage';

const config = {
  apiKey: process.env.FIREBASE_API_KEY,
  authDomain: process.env.FIREBASE_AUTH_DOMAIN,
  databaseURL: process.env.FIREBASE_DATABASE_URL,
  projectId: process.env.FIREBASE_PROJECT_ID,
  storageBucket: process.env.FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.FIREBASE_APP_ID,
};

if (!app.apps.length) {
  app.initializeApp(config);
}

const db = app.database();
const auth = app.auth();
const storage = app.storage();

// TODO
// As httpOnly cookies are to be used, do not persist any state client side.
// firebase.auth().setPersistence(firebase.auth.Auth.Persistence.NONE);

export default app;

export { db, auth, storage };
