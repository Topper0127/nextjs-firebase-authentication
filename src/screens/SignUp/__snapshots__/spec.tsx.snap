// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpPage renders 1`] = `
<section
  className="ant-layout"
>
  <header>
    <Navigation />
  </header>
  <main
    className="Layout__StyledContent-c7co8e-0 bHHvoJ ant-layout-content"
  >
    <div
      className="SignUp__SignUpPageLayout-wfg43u-0 hPcYsE"
    >
      <div
        className="ant-card SignUp__StyledCard-wfg43u-1 iRCcYv ant-card-bordered"
      >
        <div
          className="ant-card-head"
          style={Object {}}
        >
          <div
            className="ant-card-head-wrapper"
          >
            <div
              className="ant-card-head-title"
            >
              Register your account
            </div>
          </div>
        </div>
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <form
            className="ant-form ant-form-horizontal"
            onSubmit={[Function]}
          >
            <div
              className="ant-row ant-form-item Item__StyledFormItem-sc-13gzk8a-0 lhvvBo"
              style={Object {}}
            >
              <div
                className="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-10"
                style={Object {}}
              >
                <label
                  className="ant-form-item-required"
                  htmlFor="/sign-up_username"
                  onClick={[Function]}
                  title=""
                >
                  <span>
                    Personal Name
                  </span>
                </label>
              </div>
              <div
                className="ant-col ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-14"
                style={Object {}}
              >
                <div
                  className="ant-form-item-control"
                >
                  <span
                    className="ant-form-item-children"
                  >
                    <input
                      className="ant-input"
                      data-__field={
                        Object {
                          "name": "username",
                        }
                      }
                      data-__meta={
                        Object {
                          "name": "username",
                          "originalProps": Object {
                            "type": "text",
                          },
                          "ref": null,
                          "rules": Array [
                            Object {
                              "message": "Please input your name!",
                              "required": true,
                              "whitespace": true,
                            },
                          ],
                          "trigger": "onChange",
                          "validate": Array [
                            Object {
                              "rules": Array [
                                Object {
                                  "message": "Please input your name!",
                                  "required": true,
                                  "whitespace": true,
                                },
                              ],
                              "trigger": Array [
                                "onBlur",
                              ],
                            },
                          ],
                          "validateFirst": true,
                          "validateTrigger": "onBlur",
                          "valuePropName": "value",
                        }
                      }
                      id="/sign-up_username"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onKeyDown={[Function]}
                      type="text"
                      value=""
                    />
                  </span>
                </div>
              </div>
            </div>
            <div
              className="ant-row ant-form-item Item__StyledFormItem-sc-13gzk8a-0 lhvvBo"
              style={Object {}}
            >
              <div
                className="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-10"
                style={Object {}}
              >
                <label
                  className="ant-form-item-required"
                  htmlFor="/sign-up_email"
                  onClick={[Function]}
                  title="E-mail"
                >
                  E-mail
                </label>
              </div>
              <div
                className="ant-col ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-14"
                style={Object {}}
              >
                <div
                  className="ant-form-item-control"
                >
                  <span
                    className="ant-form-item-children"
                  >
                    <input
                      className="ant-input"
                      data-__field={
                        Object {
                          "name": "email",
                        }
                      }
                      data-__meta={
                        Object {
                          "name": "email",
                          "originalProps": Object {
                            "type": "text",
                          },
                          "ref": null,
                          "rules": Array [
                            Object {
                              "message": "The input is not valid E-mail!",
                              "type": "email",
                            },
                            Object {
                              "message": "Please input your E-mail!",
                              "required": true,
                            },
                          ],
                          "trigger": "onChange",
                          "validate": Array [
                            Object {
                              "rules": Array [
                                Object {
                                  "message": "The input is not valid E-mail!",
                                  "type": "email",
                                },
                                Object {
                                  "message": "Please input your E-mail!",
                                  "required": true,
                                },
                              ],
                              "trigger": Array [
                                "onBlur",
                              ],
                            },
                          ],
                          "validateFirst": true,
                          "validateTrigger": "onBlur",
                          "valuePropName": "value",
                        }
                      }
                      id="/sign-up_email"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onKeyDown={[Function]}
                      type="text"
                      value=""
                    />
                  </span>
                </div>
              </div>
            </div>
            <div
              className="ant-row ant-form-item Item__StyledFormItem-sc-13gzk8a-0 lhvvBo"
              style={Object {}}
            >
              <div
                className="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-10"
                style={Object {}}
              >
                <label
                  className="ant-form-item-required"
                  htmlFor="/sign-up_password"
                  onClick={[Function]}
                  title="Password"
                >
                  Password
                </label>
              </div>
              <div
                className="ant-col ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-14"
                style={Object {}}
              >
                <div
                  className="ant-form-item-control"
                >
                  <span
                    className="ant-form-item-children"
                  >
                    <span
                      className="ant-input-password ant-input-affix-wrapper"
                    >
                      <input
                        action="click"
                        className="ant-input"
                        data-__field={
                          Object {
                            "name": "password",
                          }
                        }
                        data-__meta={
                          Object {
                            "name": "password",
                            "originalProps": Object {
                              "action": "click",
                              "inputPrefixCls": "ant-input",
                              "prefixCls": "ant-input-password",
                              "visibilityToggle": true,
                            },
                            "ref": null,
                            "rules": Array [
                              Object {
                                "message": "Please input your password!",
                                "required": true,
                              },
                              Object {
                                "message": "Your password is too short.",
                                "min": 6,
                              },
                              Object {
                                "validator": [Function],
                              },
                            ],
                            "trigger": "onChange",
                            "validate": Array [
                              Object {
                                "rules": Array [
                                  Object {
                                    "message": "Please input your password!",
                                    "required": true,
                                  },
                                  Object {
                                    "message": "Your password is too short.",
                                    "min": 6,
                                  },
                                  Object {
                                    "validator": [Function],
                                  },
                                ],
                                "trigger": Array [
                                  "onBlur",
                                ],
                              },
                            ],
                            "validateFirst": true,
                            "validateTrigger": "onBlur",
                            "valuePropName": "value",
                          }
                        }
                        id="/sign-up_password"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        style={null}
                        type="password"
                        value=""
                      />
                      <span
                        className="ant-input-suffix"
                      >
                        <i
                          aria-label="icon: eye-invisible"
                          className="anticon anticon-eye-invisible ant-input-password-icon"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          tabIndex={-1}
                        >
                          <svg
                            aria-hidden="true"
                            className=""
                            data-icon="eye-invisible"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 0 0 0-51.5zm-63.57-320.64L836 122.88a8 8 0 0 0-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 0 0 0 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 0 0 0 11.31L155.17 889a8 8 0 0 0 11.31 0l712.15-712.12a8 8 0 0 0 0-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 0 0-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 0 1 146.2-106.69L401.31 546.2A112 112 0 0 1 396 512z"
                            />
                            <path
                              d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 0 0 227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 0 1-112 112z"
                            />
                          </svg>
                        </i>
                      </span>
                    </span>
                  </span>
                </div>
              </div>
            </div>
            <div
              className="ant-row ant-form-item Item__StyledFormItem-sc-13gzk8a-0 lhvvBo"
              style={Object {}}
            >
              <div
                className="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-10"
                style={Object {}}
              >
                <label
                  className="ant-form-item-required"
                  htmlFor="/sign-up_confirm"
                  onClick={[Function]}
                  title="Confirm Password"
                >
                  Confirm Password
                </label>
              </div>
              <div
                className="ant-col ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-14"
                style={Object {}}
              >
                <div
                  className="ant-form-item-control"
                >
                  <span
                    className="ant-form-item-children"
                  >
                    <span
                      className="ant-input-password ant-input-affix-wrapper"
                    >
                      <input
                        action="click"
                        className="ant-input"
                        data-__field={
                          Object {
                            "name": "confirm",
                          }
                        }
                        data-__meta={
                          Object {
                            "name": "confirm",
                            "originalProps": Object {
                              "action": "click",
                              "inputPrefixCls": "ant-input",
                              "onBlur": [Function],
                              "prefixCls": "ant-input-password",
                              "visibilityToggle": true,
                            },
                            "ref": null,
                            "rules": Array [
                              Object {
                                "message": "Please confirm your password!",
                                "required": true,
                              },
                              Object {
                                "message": "Your password is too short.",
                                "min": 6,
                              },
                              Object {
                                "validator": [Function],
                              },
                            ],
                            "trigger": "onChange",
                            "validate": Array [
                              Object {
                                "rules": Array [
                                  Object {
                                    "message": "Please confirm your password!",
                                    "required": true,
                                  },
                                  Object {
                                    "message": "Your password is too short.",
                                    "min": 6,
                                  },
                                  Object {
                                    "validator": [Function],
                                  },
                                ],
                                "trigger": Array [
                                  "onBlur",
                                ],
                              },
                            ],
                            "validateFirst": true,
                            "validateTrigger": "onBlur",
                            "valuePropName": "value",
                          }
                        }
                        id="/sign-up_confirm"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        style={null}
                        type="password"
                        value=""
                      />
                      <span
                        className="ant-input-suffix"
                      >
                        <i
                          aria-label="icon: eye-invisible"
                          className="anticon anticon-eye-invisible ant-input-password-icon"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          tabIndex={-1}
                        >
                          <svg
                            aria-hidden="true"
                            className=""
                            data-icon="eye-invisible"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 0 0 0-51.5zm-63.57-320.64L836 122.88a8 8 0 0 0-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 0 0 0 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 0 0 0 11.31L155.17 889a8 8 0 0 0 11.31 0l712.15-712.12a8 8 0 0 0 0-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 0 0-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 0 1 146.2-106.69L401.31 546.2A112 112 0 0 1 396 512z"
                            />
                            <path
                              d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 0 0 227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 0 1-112 112z"
                            />
                          </svg>
                        </i>
                      </span>
                    </span>
                  </span>
                </div>
              </div>
            </div>
            <div
              className="ant-row ant-form-item Item__StyledFormItem-sc-13gzk8a-0 lhvvBo"
              style={Object {}}
            >
              <div
                className="ant-col ant-form-item-control-wrapper ant-col-sm-24"
                style={Object {}}
              >
                <div
                  className="ant-form-item-control"
                >
                  <span
                    className="ant-form-item-children"
                  >
                    <button
                      className="ant-btn StretchedButton-sc-1skj5p1-0 gqwrSP ant-btn-primary"
                      onClick={[Function]}
                      type="submit"
                    >
                      <span>
                        Sign Up
                      </span>
                    </button>
                    Already have an account? 
                    <a
                      href="/sign-in"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                    >
                      Sign in!
                    </a>
                  </span>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
  <footer
    className="Layout__StyledFooter-c7co8e-1 klwLpk ant-layout-footer"
  >
    ©2020 Created by
     
    <a
      href="https://www.robinwieruch.de/"
    >
      Robin Wieruch
    </a>
  </footer>
</section>
`;
